File: 2020/16/Haskell/src/Main.hs
import           Data.Bifunctor
import           Data.HashMap.Strict            ( HashMap )
import qualified Data.HashMap.Strict           as HM
import           Data.List
import           Data.Text                      ( Text )
import qualified Data.Text                     as T
import           System.Environment
import           Text.Printf

type FieldList = [(String, [Int])]
type FieldMap = HashMap String [Int]
type Rule = (String, [(Int, Int)])
type RuleMap = HashMap String [(Int, Int)]
type Ticket = [Int]

separateInput :: String -> [[Text]]
separateInput =
  filter (not . null) . map T.lines . T.splitOn (T.pack "\n\n") . T.pack

parseRange :: Text -> (Int, Int)
parseRange =
  (\x -> (head x, last x)) . map (read . T.unpack) . T.splitOn (T.pack "-")

parseRule :: Text -> Rule
parseRule x = (i, r)
 where
  [identifier, ranges] = T.splitOn (T.pack ":") x
  i = T.unpack identifier
  r = map (parseRange . T.drop 1) . T.splitOn (T.pack "or") $ ranges

parseTicket :: Text -> Ticket
parseTicket = map (read . T.unpack) . T.splitOn (T.pack ",")

parseInput :: [[Text]] -> ([Rule], Ticket, [Ticket])
parseInput [rules, myT, nearbyTs] = (r, t, ts)
 where
  r  = map parseRule rules
  t  = parseTicket . (!! 1) $ myT
  ts = map parseTicket . drop 1 $ nearbyTs
parseInput _ = ([], [], [])

inRange :: Int -> (Int, Int) -> Bool
n `inRange` (a, b) = n >= a && n <= b

inAnyRange :: Int -> [(Int, Int)] -> Bool
n `inAnyRange` rs = foldr ((||) . (n `inRange`)) False rs

invalidEntries :: [(Int, Int)] -> Int -> Int
invalidEntries r n | n `inAnyRange` r = 0
                   | otherwise        = n

mapRules :: [Rule] -> RuleMap -> RuleMap
mapRules []       m = m
mapRules (r : rs) m = mapRules rs nextM
 where
  (identifier, ranges) = r
  nextM                = HM.insert identifier ranges m

validateEntries :: [Int] -> [(Int, Int)] -> Bool
validateEntries t rs = foldr ((&&) . (`inAnyRange` rs)) True t

validateTickets :: [Ticket] -> [(Int, Int)] -> [Ticket]
validateTickets [] _ = []
validateTickets (t : ts) rs | valid     = t : validateTickets ts rs
                            | otherwise = validateTickets ts rs
  where valid = validateEntries t rs

findField :: [String] -> [Int] -> RuleMap -> [String]
findField [] _ _ = []
findField (k : ks) x m | valid     = k : findField ks x m
                       | otherwise = findField ks x m
 where
  ranges = HM.findWithDefault [] k m
  valid  = validateEntries x ranges

insertKeys :: [String] -> Int -> FieldMap -> FieldMap
insertKeys []       _ f = f
insertKeys (x : xs) n f = insertKeys xs n nextMap
 where
  thisV = HM.findWithDefault [-1] x f
  nextV | thisV == [-1] = [n]
        | otherwise     = n : thisV
  nextMap = HM.insert x nextV f

stepFields :: [Ticket] -> Int -> FieldMap -> RuleMap -> FieldMap
stepFields t i f m | i == length (head t) = f
                   | otherwise            = stepFields t (i + 1) fieldMap m
 where
  columns  = map (!! i) t
  keys     = HM.keys m
  fields   = findField keys columns m
  fieldMap = insertKeys fields i f

fieldPreprocess :: FieldList -> FieldList
fieldPreprocess =
  map (second sort) . sortBy (\(_, x) (_, y) -> compare (length x) (length y))

processFields :: [Int] -> FieldList -> [(String, Int)]
processFields _ []       = []
processFields n (x : xs) = (i, index) : processFields nextN xs
 where
  (i, ns) = x
  nextN   = ns ++ n
  index   = head . filter (`notElem` n) $ ns

part1 :: String -> Int
part1 x = sum check
 where
  (r, _, ts) = parseInput . separateInput $ x
  ranges     = concatMap snd r
  tickets    = concat ts
  check      = map (invalidEntries ranges) tickets

assignFields :: [Rule] -> [Ticket] -> [(String, Int)]
assignFields r ts = processedFields
 where
  mapInit         = mapRules r HM.empty
  ranges          = concatMap snd r
  validTickets    = validateTickets ts ranges
  fields          = stepFields validTickets 0 HM.empty mapInit
  processedFields = processFields [] . fieldPreprocess . HM.toList $ fields

part2 :: String -> Int
part2 x = solve
 where
  (r, t, ts)      = parseInput . separateInput $ x
  processedFields = assignFields r ts
  indices =
    sort
      . map snd
      . filter ((== "departure") . head . words . fst)
      $ processedFields
  solve = product . map (t !!) $ indices

output :: String -> IO ()
output path = do
  input <- readFile path
  printf "File: %s\n" path
  printf "  Part 1: %d\n" . part1 $ input
  printf "  Part 2: %d\n" . part2 $ input

main :: IO ()
main = getArgs >>= mapM_ output

File: 2020/16/Rust/src/main.rs
mod parser;

use std::collections::{HashMap, HashSet};
use std::env;
use std::fs;

pub type Ticket = Vec<u64>;

#[derive(Debug)]
pub struct Range {
    lower: u64,
    upper: u64,
}

impl Range {
    pub fn new(lower: u64, upper: u64) -> Range {
        Range { lower, upper }
    }

    fn in_range(&self, num: &u64) -> bool {
        if (num >= &self.lower) && (num <= &self.upper) {
            return true;
        }
        false
    }
}

#[derive(Debug)]
pub struct Rule<'a> {
    id: &'a str,
    ranges: Vec<Range>,
}

impl<'a> Rule<'a> {
    pub fn new(id: &str, ranges: Vec<Range>) -> Rule {
        Rule { id, ranges }
    }
}

#[derive(Debug)]
pub struct Input<'a> {
    rules: Vec<Rule<'a>>,
    ticket: Ticket,
    nearby_tickets: Vec<Ticket>,
}

impl<'a> Input<'a> {
    pub fn new(
        rules: Vec<Rule>,
        ticket: Ticket,
        nearby_tickets: Vec<Ticket>,
    ) -> Input {
        Input {
            rules,
            ticket,
            nearby_tickets,
        }
    }
}

fn in_either_range(num: &u64, rule: &Rule) -> bool {
    for range in &rule.ranges {
        if range.in_range(num) {
            return true;
        }
    }
    false
}

fn in_any_range(num: &u64, rules: &[Rule]) -> bool {
    for rule in rules {
        if in_either_range(num, rule) {
            return true;
        }
    }
    false
}

fn validate_fields(ticket: &[u64], rules: &[Rule]) -> bool {
    ticket.iter().all(|field| in_any_range(field, rules))
}

fn get_columns(tickets: Vec<&Ticket>) -> Vec<Ticket> {
    let mut column = Vec::new();
    let mut columns = Vec::new();
    for i in 0..tickets[0].len() {
        for ticket in &tickets {
            column.push(ticket[i]);
        }
        columns.push(column.to_owned());
        column = Vec::new();
    }
    columns
}

fn step_fields<'a>(
    tickets: Vec<&'a Ticket>,
    rules: &'a [Rule],
) -> HashMap<&'a str, Vec<usize>> {
    let mut map: HashMap<&str, Vec<usize>> = HashMap::new();
    let columns = get_columns(tickets);
    for rule in rules {
        for (index, column) in columns.iter().enumerate() {
            let valid = column.iter().all(|x| in_either_range(x, rule));
            let field = rule.id;
            if valid {
                if map.get(field).is_some() {
                    map.get_mut(field).unwrap().push(index);
                } else {
                    map.insert(field, vec![index]);
                }
            }
        }
    }
    map
}

fn field_preprocess<'a>(
    fields: &'a HashMap<&str, Vec<usize>>,
) -> Vec<(&'a &'a str, &'a Vec<usize>)> {
    let mut fields = fields.iter().collect::<Vec<_>>();
    fields.sort_unstable_by(|a, b| a.1.len().cmp(&b.1.len()));
    fields
}

fn process_fields<'a>(
    fields: &'a [(&&str, &Vec<usize>)],
) -> HashMap<&'a &'a str, &'a usize> {
    let mut set = HashSet::new();
    let mut map = HashMap::new();

    for field in fields {
        let id = field.0;
        let candidates = field.1;
        for candidate in candidates {
            if set.get(candidate).is_none() {
                set.insert(candidate);
                map.insert(id, candidate);
            }
        }
    }
    map
}

fn departure_indices(map: &HashMap<&&str, &usize>) -> Vec<usize> {
    let mut result = Vec::new();
    for field in map.iter() {
        let candidate = field.0.split_whitespace().collect::<Vec<_>>();
        if let Some(id) = candidate.get(0) {
            if id == &"departure" {
                result.push(**field.1);
            }
        }
    }
    result
}

fn part_1(input: &Input) -> u64 {
    let rules = &input.rules;
    let nearby_tickets = &input.nearby_tickets;
    let mut result = 0;
    for ticket in nearby_tickets {
        for field in ticket {
            if !in_any_range(field, rules) {
                result += field;
            }
        }
    }
    result
}

fn part_2(input: &Input) -> u64 {
    let rules = &input.rules;
    let ticket = &input.ticket;
    let nearby_tickets = &input.nearby_tickets;
    let valid = nearby_tickets
        .iter()
        .filter(|ticket| validate_fields(ticket, rules))
        .collect::<Vec<_>>();
    let fields = step_fields(valid, rules);
    let fields = field_preprocess(&fields);
    let processed_fields = process_fields(&fields);
    let indices = departure_indices(&processed_fields);
    indices.iter().map(|x| ticket[*x]).product()
}

fn output(path: &str) {
    let input = fs::read_to_string(&path).unwrap();
    let parsed = parser::parse_input(&input);

    println!("File: {}", path);
    println!("  Part 1: {}", part_1(&parsed));
    println!("  Part 2: {}", part_2(&parsed));
}

fn main() {
    env::args()
        .skip(1)
        .collect::<Vec<_>>()
        .iter()
        .for_each(|x| output(x));
}
